File: sdk/typescript/src/buildCliCli.ts
import { Command } from "commander";
import { basename, resolve } from "node:path";

import { build_cli, loadAppFactory } from "./packaging.js";

export async function main(argv = process.argv): Promise<void> {
  const program = new Command()
    .name("aclip-build-cli")
    .requiredOption("--app-factory <target>", "App factory in the form './src/app.ts:createApp'.")
    .requiredOption("--entry-file <path>", "CLI entry file to bundle.")
    .option("--name <name>", "Executable name. Defaults to the ACLIP app name.")
    .option("--project-root <path>", "Project root used to resolve relative inputs.", ".")
    .option("--out-dir <path>", "Output directory for the bundled artifact and sidecar manifest.")
    .option("--package-name <name>", "npm package name to write into distribution metadata.")
    .option("--package-version <version>", "Package version to write into distribution metadata.");

  program.parse(argv);
  const options = program.opts<{
    appFactory: string;
    entryFile: string;
    name?: string;
    projectRoot: string;
    outDir?: string;
    packageName?: string;
    packageVersion?: string;
  }>();

  const projectRoot = resolve(options.projectRoot);
  const entryFile = resolve(projectRoot, options.entryFile);
  const factory = await loadAppFactory(resolveAppFactoryTarget(projectRoot, options.appFactory));
  const artifact = await build_cli({
    app: factory(),
    entryFile,
    projectRoot,
    outDir: options.outDir ? resolve(projectRoot, options.outDir) : undefined,
    executableName: options.name,
    packageName: options.packageName,
    packageVersion: options.packageVersion
  });

  process.stdout.write(`${artifact.entryPath}\n${artifact.manifestPath}\n`);
}

function resolveAppFactoryTarget(projectRoot: string, target: string): string {
  const separator = target.lastIndexOf(":");
  if (separator <= 0) {
    return target;
  }

  const modulePath = target.slice(0, separator);
  const exportName = target.slice(separator + 1);
  return `${resolve(projectRoot, modulePath)}:${exportName}`;
}

const invokedPath = process.argv[1];
if (invokedPath && /^buildCliCli\.(ts|js|cjs)$/.test(basename(invokedPath))) {
  void main().catch((error: unknown) => {
    const message = error instanceof Error ? error.message : String(error);
    process.stderr.write(`${message}\n`);
    process.exitCode = 1;
  });
}

File: sdk/python/src/aclip/build_cli_cli.py
from __future__ import annotations

import argparse
from pathlib import Path

from .packaging import build_cli, load_app_factory


def main(argv: list[str] | None = None) -> None:
    parser = argparse.ArgumentParser(prog="aclip-build-cli")
    parser.add_argument("--app-factory", required=True, help="Import target in the form 'module.path:factory_name'.")
    parser.add_argument("--entry-script", required=True, help="CLI entry script that should be frozen into the distributable.")
    parser.add_argument("--name", default=None, help="Executable name. Defaults to the ACLIP app name.")
    parser.add_argument("--project-root", default=".", help="Project root used to resolve relative inputs.")
    parser.add_argument("--source-root", default=None, help="Import root passed through to PyInstaller. Defaults to <project-root>/src when present.")
    parser.add_argument("--extra-path", dest="extra_paths", action="append", default=[], help="Additional import roots to add during packaging. Repeat for multiple paths.")
    parser.add_argument("--dist-dir", default=None, help="Output directory for the packaged executable and manifest.")
    parser.add_argument("--build-dir", default=None, help="Temporary build directory used by PyInstaller.")
    parser.add_argument("--platform", dest="platform_value", default=None, help="Platform identifier recorded in the sidecar manifest.")
    args = parser.parse_args(argv)

    project_root = Path(args.project_root).resolve()
    source_root = None
    if args.source_root is not None:
        source_root = Path(args.source_root)
        if not source_root.is_absolute():
            source_root = (project_root / source_root).resolve()
    extra_paths: list[Path] = []
    for extra_path in args.extra_paths:
        path = Path(extra_path)
        if not path.is_absolute():
            path = (project_root / path).resolve()
        extra_paths.append(path)
    entry_script = Path(args.entry_script)
    if not entry_script.is_absolute():
        entry_script = (project_root / args.entry_script).resolve()
    dist_dir = None if args.dist_dir is None else (project_root / args.dist_dir).resolve()
    build_dir = None if args.build_dir is None else (project_root / args.build_dir).resolve()

    factory = load_app_factory(args.app_factory)
    artifact = build_cli(
        app=factory(),
        entry_script=entry_script,
        project_root=project_root,
        source_root=source_root,
        extra_paths=extra_paths,
        executable_name=args.name,
        dist_dir=dist_dir,
        build_dir=build_dir,
        platform_value=args.platform_value,
    )
    print(artifact.binary_path)
    print(artifact.manifest_path)


if __name__ == "__main__":
    main()
